File: app/src/main/java/com/omarea/vboot/fragment_booster.java
package com.omarea.vboot;

import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.provider.Settings;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.ListView;
import android.widget.Switch;
import android.widget.TabHost;
import android.widget.TextView;

import com.omarea.shared.ConfigInfo;
import com.omarea.shared.ServiceHelper;
import com.omarea.shared.cmd_shellTools;
import com.omarea.ui.list_adapter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;
import static android.widget.AdapterView.OnItemClickListener;


public class fragment_booster extends Fragment {

    public fragment_booster() {
        // Required empty public constructor
    }

    View frameView;

    cmd_shellTools cmdshellTools = null;
    activity_main thisview = null;

    public static Fragment Create(activity_main thisView, cmd_shellTools cmdshellTools) {
        fragment_booster fragment = new fragment_booster();
        fragment.cmdshellTools = cmdshellTools;
        fragment.thisview = thisView;
        return fragment;
    }


    Button btn_booster_service_not_active;
    Button btn_booster_dynamicservice_not_active;
    Switch cacheclear;
    Switch dozemod;

    @Override
    public void onResume() {
        boolean serviceState = ServiceHelper.serviceIsRunning(getContext());
        btn_booster_service_not_active.setVisibility(serviceState ? GONE : VISIBLE);
        btn_booster_dynamicservice_not_active.setVisibility((serviceState && !ConfigInfo.getConfigInfo().AutoBooster) ? VISIBLE : GONE);

        super.onResume();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.layout_booster, container, false);
    }


    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        this.frameView = view;

        btn_booster_service_not_active = (Button) view.findViewById(R.id.btn_booster_service_not_active);
        btn_booster_dynamicservice_not_active = (Button) view.findViewById(R.id.btn_booster_dynamicservice_not_active);
        cacheclear = (Switch) view.findViewById(R.id.cacheclear);
        dozemod = (Switch) view.findViewById(R.id.dozemod);

        cacheclear.setChecked( ConfigInfo.getConfigInfo().AutoClearCache);
        dozemod.setChecked( ConfigInfo.getConfigInfo().UsingDozeMod);

        cacheclear.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                ConfigInfo.getConfigInfo().AutoClearCache = isChecked;
            }
        });
        dozemod.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                ConfigInfo.getConfigInfo().UsingDozeMod = isChecked;
            }
        });

        btn_booster_service_not_active.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    Intent intent = new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS);
                    startActivity(intent);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });
        btn_booster_dynamicservice_not_active.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(thisview, activity_accessibility_service_settings.class);
                startActivity(intent);
            }
        });

        final TabHost tabHost = (TabHost) view.findViewById(R.id.blacklist_tabhost);
        tabHost.setup();

        tabHost.addTab(tabHost.newTabSpec("tab_1")
                .setContent(R.id.blacklist_tab1).setIndicator("阻止后台",
                        this.getResources().getDrawable(R.drawable.check)));
        tabHost.addTab(tabHost.newTabSpec("tab_2")
                .setContent(R.id.blacklist_tab2).setIndicator("配置文件",
                        this.getResources().getDrawable(R.drawable.check)));
        tabHost.setCurrentTab(0);

        SetList(view);

        booster_blacklist = (ListView) view.findViewById(R.id.booster_blacklist);

        OnItemClickListener config_powersavelistClick = new OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, final int position, long id) {
                ToogleBlackItem(((TextView) (view.findViewById(R.id.ItemText))).getText().toString());
                CheckBox checkBox = ((CheckBox) (view.findViewById(R.id.select_state)));
                checkBox.setChecked(!checkBox.isChecked());
                System.out.print(view);
            }
        };
        booster_blacklist.setOnItemClickListener(config_powersavelistClick);
    }

    ListView booster_blacklist;

    final Handler myHandler = new Handler() {
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
        }
    };

    private void SetList(View view) {
        thisview.progressBar.setVisibility(View.VISIBLE);

        new Thread(new Runnable() {
            @Override
            public void run() {
                if (installedList == null) {
                    LoadList();
                }
                SetListData(installedList, booster_blacklist);
            }
        }).start();
    }

    void SetListData(final ArrayList<HashMap<String, Object>> dl, final ListView lv) {
        myHandler.post(new Runnable() {
            @Override
            public void run() {
                thisview.progressBar.setVisibility(View.GONE);
                lv.setAdapter(new list_adapter(getContext(), dl));
            }
        });
    }

    PackageManager packageManager;
    List<ApplicationInfo> packageInfos;
    ArrayList<HashMap<String, Object>> installedList;

    void LoadList() {
        packageManager = thisview.getPackageManager();
        packageInfos = packageManager.getInstalledApplications(0);

        installedList = new ArrayList<>();/*在数组中存放数据*/
        for (int i = 0; i < packageInfos.size(); i++) {
            if (packageInfos.get(i).sourceDir.indexOf("/system") == 0)
                continue;
            HashMap<String, Object> item = new HashMap<String, Object>();
            Drawable d = packageInfos.get(i).loadIcon(packageManager);
            item.put("icon", d);
            String pkgName = packageInfos.get(i).packageName.toLowerCase();
            item.put("select_state", false);
            for (String blitem : ConfigInfo.getConfigInfo().blacklist) {
                if (blitem.equals(pkgName)) {
                    item.put("select_state", true);
                    break;
                }
            }
            item.put("name", packageInfos.get(i).loadLabel(packageManager));
            item.put("packageName", pkgName);
            installedList.add(item);
        }
    }

    void ToogleBlackItem(String pkgName) {
        try {
            if (!ConfigInfo.getConfigInfo().blacklist.contains(pkgName)) {
                ConfigInfo.getConfigInfo().blacklist.add(pkgName);
            } else {
                ConfigInfo.getConfigInfo().blacklist.remove(pkgName);
            }
        } catch (Exception ex) {

        }
    }
}

File: app/src/main/java/com/omarea/vboot/FragmentBooster.kt
package com.omarea.vboot

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.provider.Settings
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.View.GONE
import android.view.View.VISIBLE
import android.view.ViewGroup
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import com.omarea.shared.AccessibleServiceHelper
import com.omarea.shared.AppListHelper
import com.omarea.shared.ServiceHelper
import com.omarea.shared.SpfConfig
import com.omarea.shared.model.Appinfo
import com.omarea.ui.AppListAdapter
import com.omarea.ui.OverScrollListView
import com.omarea.ui.ProgressBarDialog
import kotlinx.android.synthetic.main.layout_booster.*


class FragmentBooster : Fragment() {
    private lateinit var processBarDialog: ProgressBarDialog
    private lateinit var frameView: View
    private lateinit var blacklist: SharedPreferences
    private lateinit var editor: SharedPreferences.Editor

    override fun onResume() {
        super.onResume()

        val serviceState = AccessibleServiceHelper().serviceIsRunning(context!!)
        btn_booster_service_not_active.visibility = if (serviceState) GONE else VISIBLE
        btn_booster_dynamicservice_not_active.visibility = if (serviceState && !context!!.getSharedPreferences(SpfConfig.GLOBAL_SPF, Context.MODE_PRIVATE).getBoolean(SpfConfig.GLOBAL_SPF_AUTO_BOOSTER, false)) VISIBLE else GONE
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? =
            inflater!!.inflate(R.layout.layout_booster, container, false)


    @SuppressLint("ApplySharedPref")
    private fun bindSPF(checkBox: Switch, spf: SharedPreferences, prop: String, defValue: Boolean = false) {
        checkBox.isChecked = spf.getBoolean(prop, defValue)
        checkBox.setOnCheckedChangeListener{
            _, isChecked ->
            spf.edit().putBoolean(prop, isChecked).commit()
        }
    }

    @SuppressLint("ApplySharedPref")
    private fun bindSPF(checkBox: CheckBox, spf: SharedPreferences, prop: String, defValue: Boolean = false) {
        checkBox.isChecked = spf.getBoolean(prop, defValue)
        checkBox.setOnCheckedChangeListener{
            _, isChecked ->
            spf.edit().putBoolean(prop, isChecked).commit()
        }
    }

    @SuppressLint("ApplySharedPref", "CommitPrefEdits")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        processBarDialog = ProgressBarDialog(this.context!!)
        this.frameView = view!!
        blacklist = context!!.getSharedPreferences(SpfConfig.BOOSTER_BLACKLIST_SPF, Context.MODE_PRIVATE)
        editor = blacklist.edit()

        val spfAutoConfig = context!!.getSharedPreferences(SpfConfig.BOOSTER_SPF_CFG_SPF, Context.MODE_PRIVATE)

        bindSPF(cacheclear, spfAutoConfig, SpfConfig.BOOSTER_SPF_CFG_SPF_CLEAR_CACHE, false)
        bindSPF(dozemod, spfAutoConfig, SpfConfig.BOOSTER_SPF_CFG_SPF_DOZE_MOD, Build.VERSION.SDK_INT >= Build.VERSION_CODES.M)
        bindSPF(auto_clear_tasks, spfAutoConfig, SpfConfig.BOOSTER_SPF_CFG_SPF_CLEAR_TASKS, true)

        bindSPF(auto_switch_network_on_wifi, spfAutoConfig, SpfConfig.WIFI + SpfConfig.ON, false)
        bindSPF(auto_switch_network_on_data, spfAutoConfig, SpfConfig.DATA + SpfConfig.ON, false)
        bindSPF(auto_switch_network_on_nfc, spfAutoConfig, SpfConfig.NFC + SpfConfig.ON, false)
        bindSPF(auto_switch_network_on_gps, spfAutoConfig, SpfConfig.GPS + SpfConfig.ON, false)

        bindSPF(auto_switch_network_off_wifi, spfAutoConfig, SpfConfig.WIFI + SpfConfig.OFF, false)
        bindSPF(auto_switch_network_off_data, spfAutoConfig, SpfConfig.DATA + SpfConfig.OFF, false)
        bindSPF(auto_switch_network_off_nfc, spfAutoConfig, SpfConfig.NFC + SpfConfig.OFF, false)
        bindSPF(auto_switch_network_off_gps, spfAutoConfig, SpfConfig.GPS + SpfConfig.OFF, false)

        btn_booster_service_not_active.setOnClickListener {
            val dialog = ProgressBarDialog(context!!)
            dialog.showDialog("尝试使用ROOT权限开启服务...")
            Thread(Runnable {
                if(!AccessibleServiceHelper().startServiceUseRoot(context!!)) {
                    try {
                        myHandler.post {
                            val intent = Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS)
                            startActivity(intent)
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    } finally {
                        myHandler.post {
                            dialog.hideDialog()
                        }
                    }
                } else {
                    myHandler.post {
                        dialog.hideDialog()
                        btn_booster_service_not_active.visibility = if (AccessibleServiceHelper().serviceIsRunning(context!!)) View.GONE else View.VISIBLE
                    }
                }
            }).start()
        }
        btn_booster_dynamicservice_not_active.setOnClickListener {
            val intent = Intent(context, ActivityAccessibilitySettings::class.java)
            startActivity(intent)
        }

        val tabHost = view.findViewById(R.id.blacklist_tabhost) as TabHost
        tabHost.setup()

        tabHost.addTab(tabHost.newTabSpec("tab_1")
                .setContent(R.id.blacklist_tab1).setIndicator(context!!.getString(R.string.autobooster_tab_blacklist)))
        tabHost.addTab(tabHost.newTabSpec("tab_2")
                .setContent(R.id.blacklist_tab2).setIndicator(context!!.getString(R.string.autobooster_tab_details)))
        tabHost.addTab(tabHost.newTabSpec("tab_3")
                .setContent(R.id.blacklist_tab3).setIndicator(context!!.getString(R.string.autobooster_tab_screen)))
        tabHost.currentTab = 0

        setList()
        booster_blacklist.onItemClickListener = OnItemClickListener { _, itemView, _, _ ->
            toogleBlackItem((itemView.findViewById(R.id.ItemText) as TextView).text.toString())
            val checkBox = itemView.findViewById(R.id.select_state) as CheckBox
            checkBox.isChecked = !checkBox.isChecked
        }
    }

    internal val myHandler: Handler = Handler()

    private fun setList() {
        processBarDialog.showDialog()

        Thread(Runnable {
            if (installedList == null) {
                loadList()
            }
            setListData(installedList!!, booster_blacklist)
        }).start()
    }

    override fun onDestroy() {
        super.onDestroy()
        processBarDialog.hideDialog()
    }

    private fun setListData(dl: ArrayList<Appinfo>, lv: OverScrollListView) {
        myHandler.post {
            processBarDialog.hideDialog()
            lv.adapter = AppListAdapter(context!!, dl)
        }
    }

    private lateinit var packageManager: PackageManager
    private var installedList: ArrayList<Appinfo>? = null

    private fun loadList() {
        packageManager = context!!.packageManager
        val packageInfos = AppListHelper(context!!).getAll()
        for (item in packageInfos.iterator()) {
            item.selectState = blacklist.contains(item.packageName.toString())
        }
        installedList = packageInfos
    }

    private fun toogleBlackItem(pkgName: String) {
        if (blacklist.contains(pkgName)) {
            editor.remove(pkgName).commit()
        } else {
            editor.putBoolean(pkgName, true).commit()
        }
    }

    companion object {
        fun createPage(): Fragment {
            val fragment = FragmentBooster()
            return fragment
        }
    }
}
